### DISCUSSION TITLE
x-data alpine data error

### DISCUSSION THREAD

**Discussion by ross-oreto** [alpinejs/alpine#3968] [7. Bugs (archived)] [answered]
Following the documentation in the [Re-usable Data](https://alpinejs.dev/directives/data#re-usable-data) section of x-data, I get an error: Alpine Expression Error: testData is not defined

Using:
nodejs v20.10.0

Steps to recreate:
`npm create vite@latest alpine` -- and just use vanilla js or vanilla-ts
```
cd alpine
npm install alpinejs
```
in main.js or main.ts replace the contents with

```
import './style.css'
import Alpine from 'alpinejs'

window.Alpine = Alpine;
Alpine.data('testData', () => ({
    foo: "bar"
}));

document.querySelector<HTMLDivElement>('#app')!.innerHTML = `
  <div x-data="testData">
    <p x-text="testData.foo"></p>
  </div>
`
Alpine.start();
```
start the app on http://localhost:5174/
```
npm run dev
```

Expected result:
Alpine should populate the p tag with data "foo"

Actual result:
Alpine throws an error:
```
Expression: "testData.foo"

module.esm.js:538  Uncaught ReferenceError: testData is not defined
    at [Alpine] testData.foo (eval at safeAsyncFunction (module.esm.js:587:19), <anonymous>:3:32)
    at module.esm.js:612:21
    at tryCatch (module.esm.js:527:12)
    at module.esm.js:2880:5
    at reactiveEffect (module.esm.js:1748:16)
    at Object.effect2 [as effect] (module.esm.js:1723:5)
    at effect (module.esm.js:51:33)
    at wrappedEffect (module.esm.js:67:27)
    at Function.<anonymous> (module.esm.js:2879:3)
    at flushHandlers (module.esm.js:698:45)

```

**Comment by ekwoka** [comment, marked as answer]
Sorry to say, but this is a PEBKAC.

You reference the values on the data instance without the name of the data.

just 

```html
<p x-text="foo"></p>
```

there is no `testData` in the context for Alpine to reference at that `p` tag.

**Comment by ross-oreto** [comment]
I see, I was expecting to reference the data variable by name since it was registered with Alpine.data. 
The docs state:
"Registers a component constructor a name referenceable inside x-data expressions"

You're correct it works if I don't reference the name first.

**Reply by ekwoka** [reply]
Yes, inside the `x-data` expression, the key passed to `Alpine.data` is exposed, it is not exposed elsewhere.

And the component that the `x-data` resolves to has each property exposed.
